code/modules/atmospherics/machinery/components/turbine/turbine.dm

```diff
@@ -0,0 +1,372 @@
+	if(shaft)
+		shaft.connected = FALSE
+		shaft = null
+	turbine_parts = list()
+	UnregisterSignal(inlet, COMSIG_PARENT_QDELETING)
+	UnregisterSignal(outlet, COMSIG_PARENT_QDELETING)
+	UnregisterSignal(shaft, COMSIG_PARENT_QDELETING)
+
+/obj/machinery/power/turbine/turbine_controller/process_atmos()
+	if(!connected || !on)
+		return
+
+	if(!inlet.airs[1].gases)
+		return
+
+	var/datum/gas_mixture/input_remove = inlet.airs[1].remove_ratio(input_ratio * 0.01)
+	var/datum/gas_mixture/output = outlet.airs[1]
+
+	if(!input_remove.heat_capacity())
+		return
+
+	heat_transfer_coefficient = 1
+	efficiency_coefficient = 0
+	rpm_coefficient = 1
+
+	check_gas_composition(input_remove)
+
+	first_point.merge(input_remove)
+	var/first_point_pressure = first_point.return_pressure()
+	var/first_point_temperature = first_point.return_temperature()
+	var/datum/gas_mixture/first_remove = first_point.remove(first_point.total_moles())
+
+	second_point.merge(first_remove)
+	var/second_point_pressure = second_point.return_pressure()
+	var/second_point_temperature = second_point.temperature
+	var/heat_capacity = second_point.heat_capacity()
+
+	var/work_done = 0
+	var/delta_pressure = - (second_point_pressure - first_point_pressure)
+	if(first_point_temperature > 300 || delta_pressure > 500)
+		work_done = efficiency * second_point.total_moles() * R_IDEAL_GAS_EQUATION * first_point_temperature * log((first_point_pressure / second_point_pressure)) - rpm
+
+	rpm = (work_done ** 0.6) * 4 * rpm_coefficient
+
+	efficiency = clamp(1 - log(10, max(second_point_temperature, 1e3)) * 0.1 + efficiency_coefficient, 0, 1)
+
+	generated_power = rpm * efficiency * 10
+
+	add_avail(generated_power)
+
+	second_point.temperature = max((second_point.temperature * heat_capacity - work_done * second_point.total_moles() * 0.05 * heat_transfer_coefficient) / heat_capacity, TCMB)
+
+	var/datum/gas_mixture/second_remove = second_point.remove(second_point.total_moles())
+
+	output.merge(second_remove)
+
+	inlet.update_parents()
+	outlet.update_parents()
+
+///Check the input mixture for specific gases that will affect the performances of the turbine
+/obj/machinery/power/turbine/turbine_controller/proc/check_gas_composition(datum/gas_mixture/mix_to_check)
+	if(!mix_to_check)
+		return
+	for(var/gas_id in mix_to_check.gases)
+		switch(gas_id)
+			if(/datum/gas/water_vapor)
+				heat_transfer_coefficient += 0.2
+				efficiency_coefficient -= 0.05
+				rpm_coefficient += 0.1
+			if(/datum/gas/carbon_dioxide)
+				heat_transfer_coefficient -= 0.1
+				efficiency_coefficient += 0.1
+				rpm_coefficient += 0.03
+			if(/datum/gas/freon)
+				heat_transfer_coefficient += 0.5
+				efficiency_coefficient -= 0.1
+				rpm_coefficient += 0.3
+			if(/datum/gas/nitrous_oxide)
+				heat_transfer_coefficient += 0.05
+				efficiency_coefficient += 0.3
+				rpm_coefficient += 0.15
+			if(/datum/gas/hypernoblium)
+				heat_transfer_coefficient += 0.7
+				efficiency_coefficient += 0.2
+				rpm_coefficient += 0.3
```

Comment:
Is the plan to feed this thing fuel and have the combustion generate heat?

Reply:
The Turbine works in a closed loop, the gases inside this loop shouldn't be burning but should just act as a medium for the heat and pressure to be transfered inside the machine. These are vars that change the gas effect on the machine (they are like density and viscosity)